sql: support ALTER TABLE LOCALITY REGIONAL BY ROW to REGIONAL BY ROW

otan opened the pull request:
Resolves #59632 

Release note (sql change): Implement the ability to ALTER TABLE LOCALITY
REGIONAL BY ROW from other REGIONAL BY ROW variants.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60497)
<!-- Reviewable:end -->

ajstorm reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> All nits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60497#-:-MTLuK1oCFJnE53U3cwl:b3986iu)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 1654 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60497#-MTLmr5b3DLhk8iV5o3w:-MTLmr5b3DLhk8iV5o3x:b-gb7wbu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a418a6c26e088335dfe3d3d1291630530e60a2be/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L1654)):*
> ```
> 
> statement ok
> ALTER TABLE regional_by_row SET LOCALITY REGIONAL BY ROW AS crdb_region
> ```

Does it make sense to test the alter to REGIONAL BY ROW from REGIONAL BY ROW with a new column created in the same transaction?  Is that even supported?
___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 1771 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60497#-MTLnIixAYMg6bW8JzS5:-MTLnIixAYMg6bW8JzS6:b-3f4773) ([raw file](https://github.com/cockroachdb/cockroach/blob/a418a6c26e088335dfe3d3d1291630530e60a2be/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L1771)):*
> ```
> 
> query TT
> SHOW CREATE TABLE regional_by_row
> ```

Nit: should table name be `regional_by_row_as`?
___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 1976 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60497#-MTLnSd736R-NXAWE8sK:-MTLnSd736R-NXAWE8sL:b-316pia) ([raw file](https://github.com/cockroachdb/cockroach/blob/a418a6c26e088335dfe3d3d1291630530e60a2be/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L1976)):*
> ```
> 
> query TT
> SHOW CREATE TABLE regional_by_row_as
> ```

Nit: should table name be `regional_by_row`?
___
*[pkg/sql/alter_primary_key.go, line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60497#-MTLp3fr6nZ9puXjC9gQ:-MTLp3fr6nZ9puXjC9gR:b7m5amp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a418a6c26e088335dfe3d3d1291630530e60a2be/pkg/sql/alter_primary_key.go#L266)):*
> ```Go
> 				// the indexes if we were from a REGIONAL BY ROW.
> 				dropPartitionAllBy = true
> 			case *descpb.TableDescriptor_LocalityConfig_Global_,
> ```

Nit: can these cases be dropped?  They're empty now.
___
*[pkg/sql/alter_table_locality.go, line 237 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60497#-MTLsMph07I07kH_Cw-C:-MTLsMph07I07kH_Cw-D:b-9syu8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/a418a6c26e088335dfe3d3d1291630530e60a2be/pkg/sql/alter_table_locality.go#L237)):*
> ```Go
> 		as := n.tableDesc.LocalityConfig.GetRegionalByRow().As
> 
> 		// If the REGIONAL BY ROW (AS <col>) is exactly the same, do nothing.
> ```

Nit: One thing I've been a fan of in the past is to self-document multi-part checks like this by breaking them out into temporary variables (which the compiler should be able to optimize out anyway).  For example, what you have below could be rewritten to:

```
defaultColumnSpecified := as == nil && partColName == tree.RegionalByRowRegionNotSpecifiedName
sameAsColumnSpecified := as != nil && *as == string(partColName)
if defaultColumnSpecified || sameASColumnSpecified {
     return nil
}
```
___
*[pkg/sql/alter_table_locality.go, line 243 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60497#-MTLtRZp2a7S2Ia_XQBw:-MTLtRZp2a7S2Ia_XQBx:b-pclmjd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a418a6c26e088335dfe3d3d1291630530e60a2be/pkg/sql/alter_table_locality.go#L243)):*
> ```Go
> 		}
> 
> 		// Otherwise, signal that we have to omit the prefix to the primary key
> ```

Nit: does this work better as "Otherwise, signal that we have to omit the implicit partitioning columns when modifying the primary key"?


<!-- Sent from Reviewable.io -->

otan committed:
sql: support ALTER TABLE LOCALITY REGIONAL BY ROW to REGIONAL BY ROW

Release note (sql change): Implement the ability to ALTER TABLE LOCALITY
REGIONAL BY ROW from other REGIONAL BY ROW variants.

otan commented:
bors r=ajstorm
